banditpylib/learners/linear_bandit_learner/linucb.py

```diff
@@ -0,0 +1,122 @@
+                                        \sqrt{ \frac{\alpha  \ln(t) }{T_i(t)} } \right\}
+                """
+  def __init__(self,
+               features: List[np.ndarray],
+               delta: float,
+               lambda_reg: float,
+               name: str = None):
+    """Args:
+                    name: alias name
+                    features: feature vector of each arm in a list
+                    delta: delta
+                    lambda_reg: lambda for regularization
+            """
+    super().__init__(arm_num=len(features), name=name)
+    if delta <= 0 or delta >= 1:
+      raise ValueError('Delta is expected within (0, 1). Got %.2f.' % delta)
+
+    if lambda_reg <= 0:
+      raise ValueError('lambda_reg is expected greater than 0. Got %.2f.' %
+                       lambda_reg)
+
+    self.__delta = delta
+    self.__lambda_reg = lambda_reg
+    self.__features = features
+    self.__d = len(features[0])
+    self.__k = len(features)  # arm_nums
+
+    self.__feature_matrix = np.zeros((self.__d, self.__k))
+    for i in range(len(features)):
+      self.__feature_matrix[:, i] = features[i].reshape(-1)
+
+  def _name(self) -> str:
+    return 'linucb'
+
+  def reset(self):
+    # variables used in the algorithm
+    """
+                                d: length of each feature
+                                k: number of features/arms
+                                feature_matrix: d x k matrix of features stacked
+                                At: feature of arm played at t
+                                Xt: reward observed at t
+                                summation_AtXt: acuumulated sum of At * Xt, d x 1
+                                Vt: V matrix at time t, d x d
+                                theta_hat_t: the learners estimate of theta, d x 1
+                                """
+    self.__summation_AtXt = np.zeros((self.__d, 1))
+    self.__Vt = self.__lambda_reg * np.eye(self.__d)
+    self.__theta_hat_t = np.random.normal(0, size=(self.__d, 1))
+
+    # Current time step
+    self.__time = 1
+
+  def __LinUCB(self) -> np.ndarray:
+    """
+                                Returns:
+                                        optimistic estimate of arms' real means
+                                """
+    root_beta_t = np.sqrt(
+        self.__lambda_reg) + np.sqrt(2 * np.log(1 / self.__delta) + self.__d *
+                                     np.log(1 + (self.__time - 1) /
+                                            (self.__lambda_reg * self.__d)))
+
+    # unvectorized:
+    # for i in range(len(self.__features)):
+    #     a = self.__features[i]
+    #     a = a.reshape(-1, 1)
+    #     ucb[i] = np.dot(a.T, self.__theta_hat_t) +\
+    #         root_beta_t * np.sqrt(a.T @ np.linalg.pinv(self.__Vt) @ a)
+
+    # vectorized:
+    ucb = self.__feature_matrix.T @ self.__theta_hat_t + root_beta_t *\
```

Comment:
What LinUCB algorithm are you implementing? Shouldn't `root_beta_t` be a constant?

Reply:
the one from tor lattimore's book
beta_t is an increasing sequence